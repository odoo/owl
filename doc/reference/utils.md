# ðŸ¦‰ Utils ðŸ¦‰

Owl export a few useful utility functions, to help with common issues. Those
functions are all available in the `owl.utils` namespace.

## Content

- [`whenReady`](#whenready): executing code when DOM is ready
- [`loadJS`](#loadjs): loading script files
- [`loadFile`](#loadfile): loading a file (useful for templates)
- [`escape`](#escape): sanitizing strings
- [`debounce`](#debounce): limiting rate of function calls
- [`shallowEqual`](#shallowequal): shallow object comparison

## `whenReady`

The function `whenReady` returns a `Promise` resolved when the DOM is ready (if
not ready yet, resolved directly otherwise). If called with a callback as
argument, it executes it as soon as the DOM ready (or directly).

```js
Promise.all([loadFile("templates.xml"), owl.utils.whenReady()]).then(function ([templates]) {
  const qweb = new owl.QWeb({ templates });
  const app = new App({ qweb });
  app.mount(document.body);
});
```

or alternatively:

```js
owl.utils.whenReady(function () {
  const qweb = new owl.QWeb();
  const app = new App({ qweb });
  app.mount(document.body);
});
```

## `loadJS`

`loadJS` takes a url (string) for a javascript resource, and loads it (by adding
a script tag in the document head). It returns a promise, so the caller can
properly reacts when it is ready. Also, it is smart: it maintains a list of urls
previously loaded (or currently being loaded), and prevent doing twice the work.

For example, it is useful for lazy loading external libraries:

```js
class MyComponent extends owl.Component {
  willStart() {
    return owl.utils.loadJS("/static/libs/someLib.js");
  }
}
```

## `loadFile`

`loadFile` is a helper function to fetch a file. It simply
performs a `GET` request and returns the resulting string in a promise. The
initial usecase for this function is to load a template file. For example:

```js
async function makeEnv() {
  const templates = await owl.utils.loadFile("templates.xml");
  const qweb = new owl.QWeb({ templates });
  return { qweb };
}
```

Note that unlike `loadJS`, this function returns the content of the file as a
string. It does not add a `script` tag or any other side effect.

## `escape`

Sometimes, we need to display dynamic data (for example user-generated data) in
the user interface. If this is done by a `QWeb` template, it is not an issue:

```xml
    <div><t t-esc="user.data"/></div>
```

The `QWeb` engine will create a `div` node and add the content of the `user.data`
string as a text node, so the web browser will not parse it as html. However,
it may be a problem if this is done with some javascript code like this:

```js
class BadComponent extends Component {
  // some template with a ref to a div
  // some code ...

  mounted() {
    this.divRef.el.innerHTML = this.state.value;
  }
}
```

In this case, the content of the `div` will be parsed as html, which may inject
unwanted behaviour. To fix this, the `escape` function will simply transform a
string into an escaped version of the same string, which will be properly displayed
by the browser, but which will not be parsed as html (for example, `"<ok>"` is
escaped to the string: `"&lt;ok&gt;"`). So, the bad example above can be fixed
with the following change:

```js
this.divRef.el.innerHTML = owl.utils.escape(this.state.value);
```

## `debounce`

The `debounce` function is useful when we want to limit the number of times some
function/action is perfomed. For example, this may be useful to prevent issue
with people double clicking on a button.

It takes three arguments:

- `func` (function): this is the function that will be rate limited
- `wait` (number): this is the number of milliseconds that we want to use to
  rate limit the function `func`
- `immediate` (optional, boolean, default=false): if `immediate` is true, the
  function will be triggered immediately (leading edge of the interval). If false,
  the function will be triggered at the end (trailing edge).

It returns a function. For example:

```js
const debounce = owl.utils.debounce;
window.addEventListener("mousemove", debounce(doSomething, 100));
```

As this example shows, it is usualy useful for event handlers which are triggered
very quickly, such as `scroll` or `mousemove` events.

## `shallowEqual`

This function checks if two objects have the same values assigned to each keys:

```js
shallowEqual({ a: 1, b: 2 }, { a: 1, b: 2 }); // true
shallowEqual({ a: 1, b: 2 }, { a: 1, b: 3 }); // false
```

However, for performance reasons, it assumes that the two objects have the same
keys. If we are in a situation where this is not guaranteed, the following code
will work:

```js
const completeShallowEqual = (a, b) => shallowEqual(a, b) && shallowEqual(b, a);
```
