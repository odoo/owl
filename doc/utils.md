# ðŸ¦‰ Utils ðŸ¦‰

Owl export a few useful utility functions, to help with common issues. Those
functions are all available in the `owl.utils` namespace.

## Content

- [`whenReady`](#whenready): executing code when DOM is ready
- [`loadJS`](#loadjs): loading script files
- [`loadTemplates`](#loadtemplates): loading xml files
- [`escape`](#escape): sanitizing strings
- [`debounce`](#debounce): limiting rate of function calls
- [`shallowEqual`](#shallowequal): shallow object comparison


## `whenReady`

The function `whenReady` returns a `Promise` resolved when the DOM is ready (if
not ready yet, resolved directly otherwise). If called with a callback as
argument, it executes it as soon as the DOM ready (or directly).

```js
Promise.all([loadTemplates(), owl.utils.whenReady()]).then(function([templates]) {
  const qweb = new owl.QWeb(templates);
  const app = new App({ qweb });
  app.mount(document.body);
});
```

or alternatively:

```js
owl.utils.whenReady(function() {
  const qweb = new owl.QWeb();
  const app = new App({ qweb });
  app.mount(document.body);
});
```

## `loadJS`

`loadJS` takes a url (string) for a javascript resource, and loads it. It returns
a promise, so the caller can properly reacts when it is ready. Also, it is smart:
it maintains a list of urls previously loaded (or currently being loaded), and
prevent doing twice the work.

```js
class MyComponent extends owl.Component {
  willStart() {
    return owl.utils.loadJS("/static/libs/someLib.js");
  }
}
```

## `loadTemplates`

`loadTemplates` is a helper function to fetch a template file.  It simply
performs a `GET` request and return the resulting string in a promise. For
example:

```js
async function makeEnv() {
  const templates = await owl.utils.loadTemplates("templates.xml");
  const qweb = new owl.QWeb(templates);
  return { qweb };
}
```

## `escape`

Sometimes, we need to display dynamic data (for example user-generated data) in
the user interface.  If this is done by a `QWeb` template, it is not an issue:

```xml
    <div><t t-esc="user.data"/></div>
```

The `QWeb` engine will create a `div` node and add the content of the `user.data`
string as a text node, so the web browser will not parse it as html.  However,
it may be a problem if this is done with some javascript code like this:

```js
class BadComponent extends Component {
    // some template with a ref to a div
    // some code ...

    mounted() {
        this.divRef.el.innerHTML = this.state.value;
    }
}
```
In this case, the content of the `div` will be parsed as html, which may inject
unwanted behaviour.  To fix this, the `escape` function will simply transform a
string into an escaped version of the same string, which will be properly displayed
by the browser, but which will not be parsed as html (for example, `"<ok>"` is
escaped to the string: `"&lt;ok&gt;"`). So, the bad example above can be fixed
with the following change:

```js
    this.divRef.el.innerHTML = owl.utils.escape(this.state.value);
```


## `debounce`

The `debounce` function is useful when we want to limit the number of times some
function/action is perfomed. For example, this may be useful to prevent issue
with people double clicking on a button.

It takes three arguments:
- `func` (function): this is the function that will be rate limited
- `wait` (number): this is the number of milliseconds that we want to use to
   rate limit the function `func`
- `immediate` (optional, boolean, default=false): if `immediate` is true, the
  function will be triggered immediately (leading edge of the interval). If false,
  the function will be triggered at the end (trailing edge).

It returns a function.  For example:

```js
const debounce = owl.utils.debounce
window.addEventListener('mousemove', debounce(doSomething, 100));
```

As this example shows, it is usualy useful for event handlers which are triggered
very quickly, such as `scroll` or `mousemove` events.


## `shallowEqual`

This function checks if two objects have the same values assigned to each keys:

```js
shallowEqual({a:1, b: 2}, {a:1, b:2}); // true
shallowEqual({a:1, b: 2}, {a:1, b:3}); // false
```

However, for performance reasons, it assumes that the two objects have the same
keys. If we are in a situation where this is not guaranteed, the following code
will work:

```js
const completeShallowEqual = (a,b) => shallowEqual(a,b) && shallowEqual(b,a);
```