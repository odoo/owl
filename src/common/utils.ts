import { OwlError } from "./owl_error";

/**
 * Parses an XML string into an XML document, throwing errors on parser errors
 * instead of returning an XML document containing the parseerror.
 *
 * @param xml the string to parse
 * @returns an XML document corresponding to the content of the string
 */
export function parseXML(xml: string): XMLDocument {
  const parser = new DOMParser();
  const doc = parser.parseFromString(xml, "text/xml");
  if (doc.getElementsByTagName("parsererror").length) {
    let msg = "Invalid XML in template.";
    const parsererrorText = doc.getElementsByTagName("parsererror")[0].textContent;
    if (parsererrorText) {
      msg += "\nThe parser has produced the following error message:\n" + parsererrorText;
      const re = /\d+/g;
      const firstMatch = re.exec(parsererrorText);
      if (firstMatch) {
        const lineNumber = Number(firstMatch[0]);
        const line = xml.split("\n")[lineNumber - 1];
        const secondMatch = re.exec(parsererrorText);
        if (line && secondMatch) {
          const columnIndex = Number(secondMatch[0]) - 1;
          if (line[columnIndex]) {
            msg +=
              `\nThe error might be located at xml line ${lineNumber} column ${columnIndex}\n` +
              `${line}\n${"-".repeat(columnIndex - 1)}^`;
          }
        }
      }
    }
    throw new OwlError(msg);
  }

  return doc;
}
