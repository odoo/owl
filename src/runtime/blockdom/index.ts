export { config } from "./config";

export { toggler } from "./toggler";
export { createBlock } from "./block_compiler";
export { list } from "./list";
export { multi } from "./multi";
export { text, comment } from "./text";
export { html } from "./html";
export { createCatcher } from "./event_catcher";

export interface VNode<T = any> {
  mount(parent: HTMLElement, afterNode: Node | null): void;
  moveBeforeDOMNode(node: Node | null, parent?: HTMLElement): void;
  moveBeforeVNode(other: T | null, afterNode: Node | null): void;
  patch(other: T, withBeforeRemove: boolean): void;
  beforeRemove(): void;
  remove(): void;
  firstNode(): Node | undefined;

  el?: undefined | HTMLElement | Text;
  parentEl?: undefined | HTMLElement;
  isOnlyChild?: boolean | undefined;
  key?: any;
}

export type BDom = VNode<any>;

export function mount(vnode: VNode, fixture: HTMLElement, afterNode: Node | null = null) {
  vnode.mount(fixture, afterNode);
}

export function patch(vnode1: VNode, vnode2: VNode, withBeforeRemove: boolean = false) {
  vnode1.patch(vnode2, withBeforeRemove);
}

export function remove(vnode: VNode, withBeforeRemove: boolean = false) {
  if (withBeforeRemove) {
    vnode.beforeRemove();
  }
  vnode.remove();
}

export function withKey(vnode: VNode, key: any) {
  vnode.key = key;
  return vnode;
}
