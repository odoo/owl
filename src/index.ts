/**
 * This file is the main file packaged by rollup (see rollup.config.js).  From
 * this file, we export all public owl elements.
 *
 * Note that dynamic values, such as a date or a commit hash are added by rollup
 */
import { EventBus } from "./core/event_bus";
import { Observer } from "./core/observer";
import { QWeb } from "./qweb/index";
import { config } from "./config";
import * as _store from "./store";
import * as _utils from "./utils";
import * as _tags from "./tags";
import { AsyncRoot } from "./misc/async_root";
import { Portal } from "./misc/portal";
import * as _hooks from "./hooks";
import * as _context from "./context";
import { Link } from "./router/link";
import { RouteComponent } from "./router/route_component";
import { Router } from "./router/router";

export { Component } from "./component/component";
export { QWeb };
export { config };

export const Context = _context.Context;
export const useState = _hooks.useState;
export const core = { EventBus, Observer };
export const router = { Router, RouteComponent, Link };
export const Store = _store.Store;
export const utils = _utils;
export const tags = _tags;
export const misc = { AsyncRoot, Portal };
export const hooks = Object.assign({}, _hooks, {
  useContext: _context.useContext,
  useDispatch: _store.useDispatch,
  useGetters: _store.useGetters,
  useStore: _store.useStore,
});
export const __info__ = {};
