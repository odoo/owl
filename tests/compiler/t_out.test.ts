import {
  renderToString,
  renderToBdom,
  snapshotEverything,
  TestContext,
  makeTestFixture,
} from "../helpers";
import { mount, patch } from "../../src/blockdom";
import { createBlock } from "../../src/blockdom/index";
import { markup } from "../../src/utils";

snapshotEverything();

// -----------------------------------------------------------------------------
// t-out
// -----------------------------------------------------------------------------

describe("t-out", () => {
  let fixture: HTMLElement;

  beforeEach(() => {
    fixture = makeTestFixture();
  });

  test("literal", () => {
    const template = `<span><t t-out="'ok'"/></span>`;
    expect(renderToString(template)).toBe("<span>ok</span>");
  });

  test("literal, no outside html element", () => {
    const template = `<t t-out="'ok'"/>`;
    expect(renderToString(template)).toBe("ok");
  });

  test("variable", () => {
    const template = `<span><t t-out="var"/></span>`;
    expect(renderToString(template, { var: "ok" })).toBe("<span>ok</span>");
  });

  test("with a String class", () => {
    const template = `<span><t t-out="var"/></span>`;
    expect(renderToString(template, { var: new String("ok") })).toBe("<span>ok</span>");
  });

  test("with an extended String class", () => {
    class LoveString extends String {
      valueOf(): string {
        return `<3 ${super.valueOf()} <3`;
      }
      toString(): string {
        return this.valueOf();
      }
    }
    const template = `<span><t t-out="var"/></span>`;
    expect(renderToString(template, { var: new LoveString("ok") })).toBe(
      "<span>&lt;3 ok &lt;3</span>"
    );
  });

  test("not escaping", () => {
    const template = `<div><t t-out="var"/></div>`;
    expect(renderToString(template, { var: markup("<ok></ok>") })).toBe("<div><ok></ok></div>");
  });

  test("t-out and another sibling node", () => {
    const template = `<span><span>hello</span><t t-out="var"/></span>`;
    expect(renderToString(template, { var: markup("<ok>world</ok>") })).toBe(
      "<span><span>hello</span><ok>world</ok></span>"
    );
  });

  test("t-out with comment", () => {
    const template = `<span><t t-out="var"/></span>`;
    expect(renderToString(template, { var: markup("<p>text<!-- top secret --></p>") })).toBe(
      "<span><p>text<!-- top secret --></p></span>"
    );
  });

  test("t-out on a node with a body, as a default", () => {
    const template = `<span t-out="var">nope</span>`;
    expect(renderToString(template)).toBe("<span>nope</span>");
  });

  test("t-out with a <t/> in body", () => {
    const template = `<t t-out="var"><t></t></t>`;
    expect(renderToString(template, { var: "coucou" })).toBe("coucou");
  });

  test("t-out with just a t-set t-value in body", () => {
    const template = `<t t-out="var"><t t-set="a" t-value="1" /></t>`;
    expect(renderToString(template, { var: "coucou" })).toBe("coucou");
  });

  test("t-out on a node with a dom node in body, as a default", () => {
    const template = `<span t-out="var"><div>nope</div></span>`;
    expect(renderToString(template)).toBe("<span><div>nope</div></span>");
  });

  test("multiple calls to t-out", () => {
    const context = new TestContext();
    const sub = `
        <div>
          <t t-out="0"/>
          <div>Greeter</div>
          <t t-out="0"/>
        </div>`;

    const main = `
        <div>
          <t t-call="sub">
            <span>coucou</span>
          </t>
        </div>`;

    context.addTemplate("sub", sub);
    context.addTemplate("main", main);
    const expected =
      "<div><div><span>coucou</span><div>Greeter</div><span>coucou</span></div></div>";
    expect(context.renderToString("main")).toBe(expected);
  });

  test("t-out escaped", () => {
    const template = `<span t-out="var" />`;
    expect(renderToString(template, { var: "<div>nope</div>" })).toBe(
      "<span>&lt;div&gt;nope&lt;/div&gt;</span>"
    );
  });

  test("t-out markedup", () => {
    const template = `<span t-out="var" />`;
    expect(renderToString(template, { var: markup("<div>nope</div>") })).toBe(
      "<span><div>nope</div></span>"
    );
  });

  test("t-out switch escaped", () => {
    const template = `<span t-out="var" />`;
    const node = renderToBdom(template, { var: "<div>nope</div>" });
    mount(node, fixture);

    expect(fixture.innerHTML).toBe("<span>&lt;div&gt;nope&lt;/div&gt;</span>");
    patch(node, renderToBdom(template, { var: "<li>yep</li>" }));

    expect(fixture.innerHTML).toBe("<span>&lt;li&gt;yep&lt;/li&gt;</span>");
  });

  test("t-out switch markup", () => {
    const template = `<span t-out="var" />`;
    const node = renderToBdom(template, { var: markup("<div>nope</div>") });
    mount(node, fixture);

    expect(fixture.innerHTML).toBe("<span><div>nope</div></span>");
    patch(node, renderToBdom(template, { var: markup("<li>yep</li>") }));

    expect(fixture.innerHTML).toBe("<span><li>yep</li></span>");
  });

  test("t-out switch markup on escaped", () => {
    const template = `<span t-out="var" />`;
    const node = renderToBdom(template, { var: "<div>nope</div>" });
    mount(node, fixture);

    expect(fixture.innerHTML).toBe("<span>&lt;div&gt;nope&lt;/div&gt;</span>");
    patch(node, renderToBdom(template, { var: markup("<li>yep</li>") }));

    expect(fixture.innerHTML).toBe("<span><li>yep</li></span>");
  });

  test("t-out switch escaped on markup", () => {
    const template = `<span t-out="var" />`;
    const node = renderToBdom(template, { var: markup("<div>nope</div>") });
    mount(node, fixture);

    expect(fixture.innerHTML).toBe("<span><div>nope</div></span>");
    patch(node, renderToBdom(template, { var: "<li>yep</li>" }));

    expect(fixture.innerHTML).toBe("<span>&lt;li&gt;yep&lt;/li&gt;</span>");
  });

  test("t-out block", () => {
    const block = createBlock("<div>block</div>");
    const template = `<span t-out="bdom"/>`;
    expect(renderToString(template, { bdom: block() })).toBe("<span><div>block</div></span>");
  });

  test("t-out bdom", () => {
    const template = `<div><t t-set="var"><ol>set</ol></t><span t-out="var" /></div>`;
    expect(renderToString(template)).toBe("<div><span><ol>set</ol></span></div>");
  });

  test("t-out switch markup on bdom", () => {
    const template = `<div>
      <t t-set="bdom"><ol>set</ol></t>
      <t t-if="hasBdom">
        <span t-out="bdom" />
      </t>
      <t t-else="">
        <span t-out="var" />
      </t>
    </div>`;

    const node = renderToBdom(template, { hasBdom: true, var: markup("<li>yep</li>") });
    mount(node, fixture);

    expect(fixture.innerHTML).toBe("<div><span><ol>set</ol></span></div>");
    patch(node, renderToBdom(template, { hasBdom: false, var: markup("<li>yep</li>") }));

    expect(fixture.innerHTML).toBe("<div><span><li>yep</li></span></div>");
  });

  test("t-out 0", () => {
    const context = new TestContext();
    context.addTemplate("_basic-callee", `<div><t t-out="0" /></div>`);
    context.addTemplate("caller", `<div><t t-call="_basic-callee"><div>zero</div></t></div>`);
    const node = context.getTemplate("caller")({}, null, "");
    mount(node, fixture);
    expect(fixture.innerHTML).toBe("<div><div><div>zero</div></div></div>");
  });

  test("t-out with arbitrary object", () => {
    const template = `<div t-out="var" />`;
    const node = renderToBdom(template, { var: { someKey: "someValue" } });
    expect(() => mount(node, fixture)).toThrow();
  });

  test("t-out with arbitrary object 2", () => {
    const template = `<div t-out="var" />`;
    const node = renderToBdom(template, { var: ["someValue"] });
    expect(() => mount(node, fixture)).toThrow();
  });
});

describe("t-raw is deprecated", () => {
  test("should warn", () => {
    const template = `<div t-raw="var" />`;
    const warn = console.warn;
    const steps: string[] = [];
    console.warn = (msg: any) => steps.push(msg);

    expect(renderToString(template, { var: "<div>escaped</div>" })).toBe(
      "<div>&lt;div&gt;escaped&lt;/div&gt;</div>"
    );
    expect(steps).toEqual([
      't-raw has been deprecated in favor of t-out. If the value to render is not wrapped by the "markup" function, it will be escaped',
    ]);
    console.warn = warn;
  });

  test("t-out is actually called in t-raw's place", () => {
    const template = `<div t-raw="var" />`;
    const warn = console.warn;
    console.warn = (msg: any) => msg;

    expect(renderToString(template, { var: markup("<div>raw</div>") })).toBe(
      "<div><div>raw</div></div>"
    );
    console.warn = warn;
  });
});
