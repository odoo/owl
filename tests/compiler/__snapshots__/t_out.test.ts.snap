// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`t-out literal 1`] = `
"function anonymous(bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, component, comment } = bdom;
  let { safeOutput } = helpers;
  
  let block1 = createBlock(\`<span><block-child-0/></span>\`);
  
  return function template(ctx, node, key = \\"\\") {
    let b2 = safeOutput('ok');
    return block1([], [b2]);
  }
}"
`;

exports[`t-out literal, no outside html element 1`] = `
"function anonymous(bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, component, comment } = bdom;
  let { safeOutput } = helpers;
  
  return function template(ctx, node, key = \\"\\") {
    return safeOutput('ok');
  }
}"
`;

exports[`t-out multiple calls to t-out 1`] = `
"function anonymous(bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, component, comment } = bdom;
  let { isBoundary, zero, getTemplate } = helpers;
  const callTemplate_1 = getTemplate(\`sub\`);
  
  let block1 = createBlock(\`<div><block-child-0/></div>\`);
  let block2 = createBlock(\`<span>coucou</span>\`);
  
  return function template(ctx, node, key = \\"\\") {
    ctx = Object.create(ctx);
    ctx[isBoundary] = 1;
    let b2 = block2();
    ctx[zero] = b2;
    let b3 = callTemplate_1.call(this, ctx, node, key + \`__1\`);
    return block1([], [b3]);
  }
}"
`;

exports[`t-out multiple calls to t-out 2`] = `
"function anonymous(bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, component, comment } = bdom;
  let { zero } = helpers;
  
  let block1 = createBlock(\`<div><block-child-0/><div>Greeter</div><block-child-1/></div>\`);
  
  return function template(ctx, node, key = \\"\\") {
    let b2 = ctx[zero];
    let b3 = ctx[zero];
    return block1([], [b2, b3]);
  }
}"
`;

exports[`t-out not escaping 1`] = `
"function anonymous(bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, component, comment } = bdom;
  let { safeOutput } = helpers;
  
  let block1 = createBlock(\`<div><block-child-0/></div>\`);
  
  return function template(ctx, node, key = \\"\\") {
    let b2 = safeOutput(ctx['var']);
    return block1([], [b2]);
  }
}"
`;

exports[`t-out t-out 0 1`] = `
"function anonymous(bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, component, comment } = bdom;
  let { isBoundary, zero, getTemplate } = helpers;
  const callTemplate_1 = getTemplate(\`_basic-callee\`);
  
  let block1 = createBlock(\`<div><block-child-0/></div>\`);
  let block2 = createBlock(\`<div>zero</div>\`);
  
  return function template(ctx, node, key = \\"\\") {
    ctx = Object.create(ctx);
    ctx[isBoundary] = 1;
    let b2 = block2();
    ctx[zero] = b2;
    let b3 = callTemplate_1.call(this, ctx, node, key + \`__1\`);
    return block1([], [b3]);
  }
}"
`;

exports[`t-out t-out 0 2`] = `
"function anonymous(bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, component, comment } = bdom;
  let { zero } = helpers;
  
  let block1 = createBlock(\`<div><block-child-0/></div>\`);
  
  return function template(ctx, node, key = \\"\\") {
    let b2 = ctx[zero];
    return block1([], [b2]);
  }
}"
`;

exports[`t-out t-out and another sibling node 1`] = `
"function anonymous(bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, component, comment } = bdom;
  let { safeOutput } = helpers;
  
  let block1 = createBlock(\`<span><span>hello</span><block-child-0/></span>\`);
  
  return function template(ctx, node, key = \\"\\") {
    let b2 = safeOutput(ctx['var']);
    return block1([], [b2]);
  }
}"
`;

exports[`t-out t-out bdom 1`] = `
"function anonymous(bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, component, comment } = bdom;
  let { isBoundary, withDefault, LazyValue, safeOutput } = helpers;
  
  let block1 = createBlock(\`<div><span><block-child-0/></span></div>\`);
  let block2 = createBlock(\`<ol>set</ol>\`);
  
  function value1(ctx, node, key = \\"\\") {
    return block2();
  }
  
  return function template(ctx, node, key = \\"\\") {
    ctx = Object.create(ctx);
    ctx[isBoundary] = 1
    ctx[\`var\`] = new LazyValue(value1, ctx, node);
    let b3 = safeOutput(ctx['var']);
    return block1([], [b3]);
  }
}"
`;

exports[`t-out t-out block 1`] = `
"function anonymous(bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, component, comment } = bdom;
  let { safeOutput } = helpers;
  
  let block1 = createBlock(\`<span><block-child-0/></span>\`);
  
  return function template(ctx, node, key = \\"\\") {
    let b2 = safeOutput(ctx['bdom']);
    return block1([], [b2]);
  }
}"
`;

exports[`t-out t-out escaped 1`] = `
"function anonymous(bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, component, comment } = bdom;
  let { safeOutput } = helpers;
  
  let block1 = createBlock(\`<span><block-child-0/></span>\`);
  
  return function template(ctx, node, key = \\"\\") {
    let b2 = safeOutput(ctx['var']);
    return block1([], [b2]);
  }
}"
`;

exports[`t-out t-out markedup 1`] = `
"function anonymous(bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, component, comment } = bdom;
  let { safeOutput } = helpers;
  
  let block1 = createBlock(\`<span><block-child-0/></span>\`);
  
  return function template(ctx, node, key = \\"\\") {
    let b2 = safeOutput(ctx['var']);
    return block1([], [b2]);
  }
}"
`;

exports[`t-out t-out on a node with a body, as a default 1`] = `
"function anonymous(bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, component, comment } = bdom;
  let { safeOutput, withDefault } = helpers;
  
  let block1 = createBlock(\`<span><block-child-0/></span>\`);
  
  return function template(ctx, node, key = \\"\\") {
    let b3 = text(\`nope\`);
    let b2 = withDefault(safeOutput(ctx['var']), b3);
    return block1([], [b2]);
  }
}"
`;

exports[`t-out t-out on a node with a dom node in body, as a default 1`] = `
"function anonymous(bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, component, comment } = bdom;
  let { safeOutput, withDefault } = helpers;
  
  let block1 = createBlock(\`<span><block-child-0/></span>\`);
  let block3 = createBlock(\`<div>nope</div>\`);
  
  return function template(ctx, node, key = \\"\\") {
    let b3 = block3();
    let b2 = withDefault(safeOutput(ctx['var']), b3);
    return block1([], [b2]);
  }
}"
`;

exports[`t-out t-out switch escaped 1`] = `
"function anonymous(bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, component, comment } = bdom;
  let { safeOutput } = helpers;
  
  let block1 = createBlock(\`<span><block-child-0/></span>\`);
  
  return function template(ctx, node, key = \\"\\") {
    let b2 = safeOutput(ctx['var']);
    return block1([], [b2]);
  }
}"
`;

exports[`t-out t-out switch escaped on markup 1`] = `
"function anonymous(bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, component, comment } = bdom;
  let { safeOutput } = helpers;
  
  let block1 = createBlock(\`<span><block-child-0/></span>\`);
  
  return function template(ctx, node, key = \\"\\") {
    let b2 = safeOutput(ctx['var']);
    return block1([], [b2]);
  }
}"
`;

exports[`t-out t-out switch markup 1`] = `
"function anonymous(bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, component, comment } = bdom;
  let { safeOutput } = helpers;
  
  let block1 = createBlock(\`<span><block-child-0/></span>\`);
  
  return function template(ctx, node, key = \\"\\") {
    let b2 = safeOutput(ctx['var']);
    return block1([], [b2]);
  }
}"
`;

exports[`t-out t-out switch markup on bdom 1`] = `
"function anonymous(bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, component, comment } = bdom;
  let { isBoundary, withDefault, LazyValue, safeOutput } = helpers;
  
  let block1 = createBlock(\`<div><block-child-0/><block-child-1/></div>\`);
  let block2 = createBlock(\`<ol>set</ol>\`);
  let block3 = createBlock(\`<span><block-child-0/></span>\`);
  let block5 = createBlock(\`<span><block-child-0/></span>\`);
  
  function value1(ctx, node, key = \\"\\") {
    return block2();
  }
  
  return function template(ctx, node, key = \\"\\") {
    ctx = Object.create(ctx);
    ctx[isBoundary] = 1
    let b3,b5;
    ctx[\`bdom\`] = new LazyValue(value1, ctx, node);
    if (ctx['hasBdom']) {
      let b4 = safeOutput(ctx['bdom']);
      b3 = block3([], [b4]);
    } else {
      let b6 = safeOutput(ctx['var']);
      b5 = block5([], [b6]);
    }
    return block1([], [b3, b5]);
  }
}"
`;

exports[`t-out t-out switch markup on escaped 1`] = `
"function anonymous(bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, component, comment } = bdom;
  let { safeOutput } = helpers;
  
  let block1 = createBlock(\`<span><block-child-0/></span>\`);
  
  return function template(ctx, node, key = \\"\\") {
    let b2 = safeOutput(ctx['var']);
    return block1([], [b2]);
  }
}"
`;

exports[`t-out t-out with a <t/> in body 1`] = `
"function anonymous(bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, component, comment } = bdom;
  let { safeOutput } = helpers;
  
  return function template(ctx, node, key = \\"\\") {
    return safeOutput(ctx['var']);
  }
}"
`;

exports[`t-out t-out with arbitrary object 1`] = `
"function anonymous(bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, component, comment } = bdom;
  let { safeOutput } = helpers;
  
  let block1 = createBlock(\`<div><block-child-0/></div>\`);
  
  return function template(ctx, node, key = \\"\\") {
    let b2 = safeOutput(ctx['var']);
    return block1([], [b2]);
  }
}"
`;

exports[`t-out t-out with arbitrary object 2 1`] = `
"function anonymous(bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, component, comment } = bdom;
  let { safeOutput } = helpers;
  
  let block1 = createBlock(\`<div><block-child-0/></div>\`);
  
  return function template(ctx, node, key = \\"\\") {
    let b2 = safeOutput(ctx['var']);
    return block1([], [b2]);
  }
}"
`;

exports[`t-out t-out with comment 1`] = `
"function anonymous(bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, component, comment } = bdom;
  let { safeOutput } = helpers;
  
  let block1 = createBlock(\`<span><block-child-0/></span>\`);
  
  return function template(ctx, node, key = \\"\\") {
    let b2 = safeOutput(ctx['var']);
    return block1([], [b2]);
  }
}"
`;

exports[`t-out t-out with just a t-set t-value in body 1`] = `
"function anonymous(bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, component, comment } = bdom;
  let { safeOutput } = helpers;
  
  return function template(ctx, node, key = \\"\\") {
    return safeOutput(ctx['var']);
  }
}"
`;

exports[`t-out variable 1`] = `
"function anonymous(bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, component, comment } = bdom;
  let { safeOutput } = helpers;
  
  let block1 = createBlock(\`<span><block-child-0/></span>\`);
  
  return function template(ctx, node, key = \\"\\") {
    let b2 = safeOutput(ctx['var']);
    return block1([], [b2]);
  }
}"
`;

exports[`t-raw is deprecated should warn 1`] = `
"function anonymous(bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, component, comment } = bdom;
  let { safeOutput } = helpers;
  
  let block1 = createBlock(\`<div><block-child-0/></div>\`);
  
  return function template(ctx, node, key = \\"\\") {
    let b2 = safeOutput(ctx['var']);
    return block1([], [b2]);
  }
}"
`;

exports[`t-raw is deprecated t-out is actually called in t-raw's place 1`] = `
"function anonymous(bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, component, comment } = bdom;
  let { safeOutput } = helpers;
  
  let block1 = createBlock(\`<div><block-child-0/></div>\`);
  
  return function template(ctx, node, key = \\"\\") {
    let b2 = safeOutput(ctx['var']);
    return block1([], [b2]);
  }
}"
`;
